Floofbot/Modules/UserAssignableRoleCommands.cs

```diff
@@ -0,0 +1,257 @@
+﻿using System;
+using Discord;
+using System.Threading.Tasks;
+using Discord.Commands;
+using System.Text.RegularExpressions;
+using Floofbot.Services.Repository;
+using Floofbot.Services.Repository.Models;
+using System.Linq;
+using Discord.Addons.Interactive;
+using Discord.WebSocket;
+using Serilog;
+
+namespace Floofbot.Modules
+{
+    [Summary("Join roles that the administrator have allowed you to")]
+    public class UserAssignableRoleCommands : InteractiveBase
+    {
+        private FloofDataContext _floofDb;
+        private readonly Discord.Color EMBED_COLOUR = new Discord.Color((uint)new Random().Next(0x1000000));
+        public UserAssignableRoleCommands(FloofDataContext floofDb)
+        {
+            _floofDb = floofDb;
+        }
+
+        private SocketRole GetRole(string roleName, SocketGuild guild)
+        {
+            foreach (SocketRole r in guild.Roles)
+            {
+                if (r.Name.ToLower() == roleName)
+                    return r;
+            }
+            return null;
+        }
+        private bool CheckRoleEntryExists(SocketRole role, SocketGuild guild)
+        {
+            // checks if a word exists in the filter db
+            bool isRoleInDb = _floofDb.UserAssignableRoles.AsQueryable().Where(r => r.ServerId == guild.Id).Where(r => r.RoleId == role.Id).Any();
+            return isRoleInDb;
+        }
+
+        [Summary("Join a role that the server administrators allow you to")]
+        [Command("iam")]
+        public async Task IAm([Summary("role name")][Remainder]string roleName)
+        {
+            if (roleName == null)
+            {
+                await Context.Channel.SendMessageAsync("", false, new EmbedBuilder { Description = $"💾 Usage: `iam [rolename]`", Color = EMBED_COLOUR }.Build());
+                return;
+            }
+
+            // find the role by iterating over guild roles
+            SocketRole role = GetRole(roleName.ToLower(), Context.Guild);
+            if (role == null)
+            {
+                await Context.Channel.SendMessageAsync("Unable to find a role with that name.");
+                return;
+            }
+
+            // check that they are actually allowed to join the role
+            bool roleExistsInDb = CheckRoleEntryExists(role, Context.Guild);
```

Comment:
So we just have to make sure the roles in the table are all user-assignable ones?

Reply:
yes all roles in the database are ones that users are allowed to join